README
TO BET TYPE IN THE URL http://localhost:8001/api/

**WITH EITHER THE EXTENTION**
/single/bet/amount
/multiple/bets


/* Types of bet
	Number 		(0-36)
	Red or Black	red/black
	Even or Odd	even/odd 
	Columns		c1, c2, c3
	Twelves		t1, t2, t3	
	Eighteen	e1, e2		
*/

EXAMPLES

Betting 5 on red
http://localhost:8001/api/single/red/5

Betting 5 on 0, 5 on red and 5 on 12
http://localhost:8001/api/multiple/0-5-red-5-12-5

Betting 5 on column1
http://localhost:8001/api/single/c1/5



history.txt
24,25,2,20,26,20,32
,26,8,20,34,22,26,6,36,25,14,15,34,10,5,2,27,8,0,24,27,5,20,26,28,35,12,27,6,15,6,32,18,4,30,18,12,20,20,15,26,28,14,31,34,6,32,24,3,14,12,12,23,16,3,4,10,13,25,12,2,26,22,12,24,30,8,29,23,4,22,27,15,1,1,16,10,14,8,1,15,5,1,24,0,2,9,4

spin.js

var rn = require('random-number');
var fs = require('fs')

var gen = rn.generator({
	min: 0,
	max: 36,
	integer: true
});

function spin(callback) {
	var value = gen();
	var colour = dictionary[value];

	fs.appendFile('history.txt', "," + value, function (err) {
	  if (err) throw err;
	});
	callback([value, colour]);
}

module.exports = {
	spin: spin
};

var dictionary = {
	0: "green",
	1:"red",
	3:"red",
	5:"red",
	7:"red",
	9:"red",
	12:"red",
	14:"red",
	16:"red",
	18:"red",
	19:"red",
	21:"red",
	23:"red",
	25:"red",
	27:"red",
	30:"red",
	32:"red",
	34:"red",
	36:"red",
	2:"black",
	4:"black",
	6:"black",
	8:"black",
	10:"black",
	11:"black",
	13:"black",
	15:"black",
	17:"black",
	20:"black",
	22:"black",
	24:"black",
	26:"black",
	28:"black",
	29:"black",
	31:"black",
	33:"black",
	35:"black",
}


server.js

var express    = require('express');        // call express
var app        = express();                 // define our app using express
var bodyParser = require('body-parser');
var rn = require('random-number');
var spinner = require('./spin');
var winnings = require('./winnings');
var account = require('./accounts');
var fs = require('fs')

// configure app to use bodyParser()
// this will let us get the data from a POST
app.use(bodyParser.urlencoded({ extended: true }));
app.use(bodyParser.json());

var port = process.env.PORT || 8002;        // set our port

// ROUTES FOR OUR API
var router = express.Router();              // get an instance of the express Router

router.get('/', function(req, res) {
    res.json({ message: 'Welcome to the roulette api' });   
});

//Get the history of the previous spins
router.get('/history', function(req,res) {
	showData = fs.readFile('history.txt', 'utf8', function(err, data) {
		if(err) return console.log(err);
		res.send(data);
	});
});

//Test route
router.get('/test', function(req,res) {
	
});

// accessed via http://localhost:8001/api/single/bet/amount/account
// example http://localhost:8001/api/single/red/5/001
router.get('/single/:bet/:amount/:account/', function(req, res) {
	account.canBet(req.params.account, req.params.amount, function(isSufficient) {
		if (isSufficient) {
			spinner.spin(function(result) {
				console.log(req.params);
				console.log(result);
				winnings.returnWinnings(req.params.bet, result, req.params.amount, function(winnings) {
					account.updateBalance(req.params.account, (winnings - req.params.amount), function(calls) {
						res.send("Balance updated for account " + req.params.account + ", you won " + winnings + " betting " + req.params.bet + ". The result was (" +  result + ")");
					});
				});
			});
		}
		else {
			res.send("Insufficient funds");
		}
	});
});

// accessed via http://localhost:8001/api/multiple/bets
// accessed via http://localhost:8001/api/multiple/0-5-red-5-12-5 // Bets 5 on 0, red and 12
router.get('/multiple/:bet/:account/', function(req, res) {
	var allBets = req.params.bet.split("-");
	//totalBets returns the total amount bet across all bets
	totalBets(req.params.bet.split("-"), function(totalBet) {
		/*Check the total bet amount against the amount in the their account
		  Returns isSufficient either true if able to bet or false if unable to bet */
		account.canBet(req.params.account, totalBet, function(isSufficient) {
			console.log("This account was able to bet: " + isSufficient);
			if(isSufficient) {
				console.log("bets: " + allBets);
				spinner.spin(function(result) {
					console.log("result: " + result);
					doBets(allBets, result, function(moneyReturned){
						account.updateBalance(req.params.account, (moneyReturned - totalBet), function(calls) {
							res.send("Balance updated for account " + req.params.account + ", you won " + moneyReturned + " betting " + totalBet + ". The result was (" +  result + ")");
						});
					});
				});
			}
			else {
				res.send("Insufficient funds");
			}
		});
	});
});

/*--- Runs through the bets when multiple bets are placed  --*/
function doBets(bets, result, callback) {
		var payable = 0;
		for (var x = 0; x < (bets.length); x += 2) {	
			winnings.returnWinnings(bets[x+1], result, bets[x], function(wino) {
				payable += wino;
			});
			if (x == bets.length-2) {
				console.log("winnings: " + payable);
				callback(payable);
			}
		}
}

function totalBets(bets, callback) {
	var betTotal = 0;
	for (var x = 0; x < (bets.length); x+=2) {
		betTotal += Number.parseInt(bets[x]);
		if (x == bets.length-2) {
			callback(betTotal);
		}
	}
}

// all of our routes will be prefixed with /api
app.use('/api', router);

app.listen(port);
console.log('Welcome to the roulette API on port ' + port);



accounts.js
//This code is responsiuble for accessing users accounts
var fs = require('fs')
var bodyParser = require('body-parser')

//Functionality to come later
function verifyUser(userName, password) {

}

//Returns a bool
function canBet(userID, betAmount, callback) {
	getBalance(userID, function(balance) {
			if (betAmount > balance.balance) {
				callback(false);
			}
			else if (betAmount <= balance.balance) {
				callback(true);
			}
			else {
				callback(null);
			}	
	});
}

//Returns the latest balance after betting
function updateBalance(userID, amount, callback) {
	fs.readFile('accounts.json', 'utf8', function(err, data){
	    if (err){
	        console.log(err);
	    } 
	    else {
		    obj = JSON.parse(data); 
		    obj[userID].balance += amount;
		    json = JSON.stringify(obj); 
		    fs.writeFile('accounts.json', json, 'utf8', callback);
		}
	});
}

//Returns a float of their balance
function getBalance(userID, callback) {
	fs.readFile('accounts.json', 'utf8', function(err, data) {
		if(err) return console.log(err);
		callback((JSON.parse(data))[userID]);
	});
}

module.exports = {
	canBet : canBet,
	updateBalance: updateBalance,
	getBalance: getBalance
}

winnings.js
/* Types of bet
	Number 		(0-36)
	Red or Black	red/black
	Even or Odd	even/odd 
	Columns		c1, c2, c3
	Twelves		t1, t2, t3	
	Eighteen	e1, e2		
*/

var c1 = [1,4,7,10,13,16,19,22,25,28,31,34]
var c2 = [2,5,8,11,14,17,20,23,26,29,32,35]
var c3 = [3,6,9,12,15,18,21,24,27,30,33,36]
	
function returnWinnings(bet, result, amount, callback) {
	if (bet >= 0 || bet < 37 ) {
		if (bet == result[0]) {
			callback(amount * 36);
		}
		else {callback(0);}
	}
	else {
		switch (bet) {
			case "red":
				redBlack(bet,result,amount,callback);
				break;
			case "black":
				redBlack(bet,result,amount,callback);
				break;
			case "odd":
				oddEven(bet,result,amount,callback);
				break;
			case "even":
				oddEven(bet,result,amount,callback);
				break;
			case "c1":
				isMatch(c1, result[0], amount, callback);
				break;
			case "c2":
				isMatch(c2, result[0], amount, callback);
				break;
			case "c3":
				isMatch(c3, result[0], amount, callback);
				break;
			case "t1":
				t1(result[0], amount, callback);
				break;
			case "t2":
				t2(result[0], amount, callback);
				break;
			case "t3":
				t3(result[0], amount, callback);
				break;
			case "e1":
				e1(result[0], amount, callback);
				break;
			case "e2":
				e2(result[0], amount, callback);
				break;
			}
	}
}

//Determines if red or black bet wins
function redBlack (bet, result, amount, callback) {
	if (bet == result[1]) {
		callback(amount * 2);
	}
	else {
		callback(0);
	}
}

//Determines if odd or even bet wins
function oddEven (bet, result, amount, callback) {
	if (result[0] % 2 == 0 && bet == "even") {
		callback(amount *2);
	}
	else if (result[0] % 2 == 1 && bet == "odd") {
		callback(amount *2);
	}
	else {
		callback(0);
	}
}

function t1(result, amount, callback) {
	if (result < 13) {
		callback(amount * 3);
	}
	else {
		callback(0);
	}
}

function t2(result, amount, callback) {
	if (result > 12 && result < 25) {
		callback(amount * 3);
	}
	else {
		callback(0);
	}	
}

function t3(result, amount, callback) {
	if (result > 24) {
		callback(amount * 3);
	}	
	else {
		callback(0);
	}
}

function e1(result, amount, callback) {
	if (result < 19) {
		callback(amount * 2);		
	}
	else {
		callback(0);
	}
}

function e2(result, amount, callback) {
	if (result > 18) {
		callback(amount * 2);
	}
	else {
		callback(0);
	}
}
	

function isMatch (bet, result, amount, callback) {
	for (var x = 0; x < c1.length; x++) {
		if (result == bet[x]) {
			return callback(amount * 3);
		}
		else if(x+1 == c1.length) {
			callback(0);
		}	
	}
}


module.exports = {
	returnWinnings : returnWinnings
}

